xla/xla.bzl

```diff
@@ -71,10 +71,62 @@ def xla_cc_binary(deps = [], copts = tsl_copts(), **kwargs):
     native.cc_binary(deps = deps + _XLA_SHARED_OBJECT_SENSITIVE_DEPS, copts = copts, **kwargs)
 
 def xla_cc_test(name, deps = [], use_gpu = False, **kwargs):
-    # Need to do it this way so that `tf_exec_properties` can read tags.
-    _tags = kwargs.get("tags", [])
-    kwargs["tags"] = _tags + tf_gpu_tests_tags() if use_gpu else _tags
+    """ Generates different cc_tests for generic nvidia and amd gpu backends
 
+    The name of the cc_tests are the provided name argument with the backend name
+    appended. For nvidia case if name parameter is foo_test then test is expanded to
+    foo_test_gpu_any and for AMD it would be foo_test_amd_gpu_any
+
+    Each generated cc_test target has a tag indicating which backend the test is
+    for. This tag is of the form requires-gpu-nvidia and/or requires-gpu-amd
+
+    Args:
+      name: Name of the target.
+      deps: Dependencies of the target.
+      use_gpu: True if gpu backend
+      **kwargs: Additional keyword arguments to pass to native.cc_test.
+    """
+    if use_gpu:
+        test_names = []
+        backends = ['gpu_any', 'amd_gpu_any']
+        for backend in backends:
+            test_name = "%s_%s" % (name, backend)
+            _tags = kwargs.get('tags', [])
+            _copts = kwargs.get('copts', [])
+            if backend == 'gpu_any':
+                this_backend_tags = tf_gpu_tests_tags()
+            elif backend == 'amd_gpu_any':
+                if 'requires-gpu-nvidia' in _tags:
+                    _tags.remove('requires-gpu-nvidia')
+                this_backend_tags = ['requires-gpu-amd']
+            else:
+                fail("Unknown backend %s" % backend)
+            this_backend_copts = ["-DXLA_TEST_BACKEND_GPU=1"]
+            kwargs['tags'] = _tags + this_backend_tags
+            kwargs['copts'] = _copts + ["-DXLA_TEST_BACKEND_%s=1" % backend.upper()] + this_backend_copts
+
+            xla_cc_test_backend(
+                name = test_name,
+                deps = deps,
+                **kwargs
+            )
+
+            test_names.append(test_name)
+
+        # Notably, a test_suite with `tests = []` is not empty:
+        # https://bazel.build/reference/be/general#test_suite_args and the default
+        # `tests = []` behavior doesn't respect `--build_tag_filters` due to
+        # b/317293391. For this reason, if we would create an empty `test_suite`,
+        # instead create a `cc_test` with no srcs that links against `main` to have
+        # more predictable behavior that avoids bugs.
+        if test_names:
+            native.test_suite(name = name, tags = kwargs['tags'], tests = test_names)
+        else:
+            native.cc_test(name = name, deps = ["@tsl//tsl/platform:test_main"])
```

Comment:
This seems to be replicating a lot of logic that we also have in `xla_test`. I'm wondering if we could remove the `use_gpu` flag from `xla_cc_test` entirely and make all the users of this flag use the `xla_test` macro instead. Is that something you have looked at by chance?

Reply:
I thought about it but that is almost 550+ occurrence of  `xla_cc_test`. It would be better to do it as a separate cleanup commit to ease to reviewing process.

Reply:
70 of them with use_gpu = True mostly in 3 or 4 directories, which can be added with this change.

Reply:
Hey, I went ahead and did the change in #13092. Once that has landed you should be able to rebase (hopefully rather easily).

Reply:
The change just went in.